python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -388,6 +399,43 @@ def terminate_scheduled_nodes(self):
         self.nodes_to_terminate = []
         self.update_worker_list()
 
+    def drain_nodes_via_gcs(self):
+        node_ips = [
+            self.provider.internal_ip(node_id)
+            for node_id in self.nodes_to_terminate
+        ]
+        raylet_ids_to_drain = {
+            self.load_metrics.raylet_id_by_ip[ip]
+            for ip in node_ips
+        }
+
+        logger.info(f"Draining raylets with ids {raylet_ids_to_drain}.")
+        request = gcs_service_pb2.DrainNodeRequest(drain_node_data=[
+            gcs_service_pb2.DrainNodeData(node_id=raylet_id)
+            for raylet_id in raylet_ids_to_drain
+        ])
+        try:
+            response = self.gcs_node_info_stub.DrainNode(request)
+            drained_raylet_ids = {
+                status_item.node_id
+                for status_item in response.drain_node_status
+            }
+            failed_to_drain = raylet_ids_to_drain - drained_raylet_ids
+            if failed_to_drain:
+                logger.error(
+                    f"Failed to drain raylets with ids {failed_to_drain}.")
```

Comment:
does this mean that the request will not crash if an IP is provided but raylet does not see it?

Reply:
I'm not sure if with the current implementation it's possible for only some nodes to drain successfully, others not, and for the request to return successfully.
It's possible that the logic in these lines is unnecessary. 

wdyt @rkooo567 ?

Reply:
It is guaranteed that all nodes are drained for gcs perspective when the reply comes in. It is not guaranteed all nodes are actually informed that they are drained.

Reply:
What happens if a node id is sent here but gcs does not know about it? does it crash?

Reply:
I guesss my question is would we ever be inside this failed_to_drain condition or when it happens we will already be in an exception?

Reply:
So, as long as the RPC is replied, GCS definitely knows all nodes are drained. 

So you are right about the second sentence. When you are in this condition, there's already an exception

Reply:
Super, will simplify by removing the logic.

Reply:
What I would do is probably keep some logger.error here in case this happens but also handle the exception properly

Reply:
Will leave as is for now, then.
If there's an exception (besides unimplemented) it will be logged in the `except Exception` part -- we can refine the handling later if needed.